design/KEPS/multicluster/proposal.md

```diff
@@ -0,0 +1,160 @@
+
+The API routes to a particular s/w/a based on the namespace/name e.g. `/api/namespaces/{namespace}/workloads/{workload}`. These namespace specific routes will stay the same but they will include results from all clusters. e.g. given a `bookinfo/reivews` service in cluster A and cluster B, `/api/namespaces/bookinfo/services/reviews` would return info about a reviews service from both cluster A and B.
+
+It will be up to the business layer to aggregate all of the info across clusters. Fetching workloads and services will need to be done inside the business layer for every cluster Kiali is connected to. Since details of workloads/services can vary based on the cluster, e.g. the `reviews.bookinfo` Service in cluster A may have different labels than on cluster B and Kiali needs to show those details, the Kiali details models will need to distinguish between the `Service` objects on cluster A vs. cluster B. Istio config will also need a way to uniquely identify istio config across clusters. Istio configuration is not shared across clusters and is specific to the cluster it resides in.
+
+Some services/workloads may span across clusters but some may not. Business layer services can optionally accept a list of specific clusters to query against for a given s/w/a as an optimization. The business services will need to handle this aggregation properly and cannot assume that s/w/a will exist in all clusters or only in a single cluster.
+
+## UI/UX
+
+This proposal focuses mostly on the backend changes necessary to enable a multi-cluster deployment of Kiali however some UI elements can and should change with a multi-cluster deployment. Exact details of UI changes will be left to future issues.
+
+## Configuration
+
+To deploy a multi-cluster Kiali, users are expected to configure Kiali with access to:
+
+- Centralized prometheus
+- Centralized Jaeger
+- 1..N Kube API servers where N is the number of kube clusters in the mesh. Kiali will require kube api credentials to communicate with each of these kube api servers.
+
+Access to the Kube API servers can be granted by:
+
+1. Creating a service account in the target cluster.
+2. Granting proper permissions for this service account.
+3. Creating a secret with the credentials for this service account in the cluster/namespace where Kiali is deployed.
+
+This process will need to be repeated for every new cluster added. Adding or removing clusters will also require restarting the Kiali pod in order for the cache watches to pick up the new kube config. Adding/removing kube clusters without having to restart Kiali can be an improvement made later.
+
+## Operator
+
+The Kiali operator will still be used for configuring Kiali itself however, it's assumed that the operator won't have access to other kube clusters other than where Kiali is deployed and therefore configuring multicluster access for Kiali will be a manual operation. The operator will remain unchanged in a multi-cluster deployment.
```

Comment:
The credentials will be a kubeconfig file that fully describes how to connect to the cluster. That will contain details on how to connect to the cluster and metadata about the cluster. We shouldn't need to tie the credential file name to the cluster name. We can provide a script to help generate this.

You will need to mount them into the kiali pod but the naming shouldn't matter so much for identifying the cluster.

Reply:
Which type of secret are we assuming the user will create? Types are: https://kubernetes.io/docs/concepts/configuration/secret/#secret-types

I am not sure how it will work if we don't identify which credentials belong to which cluster. For example, if the graph is showing 2 nodes, one from cluster-1 and one from cluster-2, and the user wants to see details for the node for cluster-2, how does the Kiali server know which credentials to use to connect to cluster-2 to get data for that node?

Reply:
See PRs related to: https://github.com/kiali/kiali/issues/5738